### PR TITLE
Added the File Uploader UI Common Component

### PR DESCRIPTION
**Pull request by roshan-bisht-lab49**
Signed-off-by: Roshan Singh Bisht <roshan.bisht@lab49.com>

Screenshot 2022-12-21 at 2 38 44 PM
Screenshot 2022-12-21 at 2 39 26 PM

### CODE REVIEW

File: src/dex-ui-components/base/Inputs/FileUploader.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import { useDropzone } from "react-dropzone";
+import { Card, Text } from "@chakra-ui/react";
+import { Color } from "../../../dex-ui-components";
+interface File extends Blob {
+  readonly lastModified: number;
+  readonly name: string;
+  readonly webkitRelativePath: string;
+}
+
+const defaultFileUploaderStyles = {
```
Comment (damiano-melcarne-lab49) [inline src/dex-ui-components/base/Inputs/FileUploader.tsx:11]: To stick with the ChakraUI styling config pattern, you can add these styles as a `variant` in the `CardStyles` object in **dex-ui-components/base/Card/styles.ts**.

`
export const CardStyles: ComponentStyleConfig = {
...
variants: {
"file-uploader": {
...styles
}
}
...
}
`

`
<Card {...getRootProps()} variant="file-uploader" style={{ ...fileUploaderStyles }}>
`

File: src/dex-ui-components/base/Inputs/FileUploader.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  });
+
+  return (
+    <Card {...getRootProps()} style={{ ...defaultFileUploaderStyles, ...fileUploaderStyles }}>
+      <input {...getInputProps()} />
+      {isDragActive ? (
+        <Text color={Color.Blue_03}>{onHoverTitle}</Text>
+      ) : (
+        <>
+          <Text color={Color.Blue_03}>{title}</Text>
```
Comment (damiano-melcarne-lab49) [inline src/dex-ui-components/base/Inputs/FileUploader.tsx:67]: Use the `textStyle` variants to define text attributes. You can then override attributes like `color` using inline styles (like you are already doing). 

It looks like the wireframes specified this as `body-02` text which maps to `b2` in our code.
Screen Shot 2022-12-21 at 11 50 15 AM

`<Text textStyle="b2" color={Color.Blue_03}>{title}</Text>`

The list of available custom text variants extracted from the wireframes is in the **dex-ui-components/base/Text/styles/TextStyles.ts** file.

Note that Typography defined in the wireframes should but may not always map one-to-one with the text variants described in our code. If you run into any discrepancies, you can reach out to myself or Anna.

Comment (damiano-melcarne-lab49) [reply src/dex-ui-components/base/Inputs/FileUploader.tsx:67]: Also, it looks like the wireframes specify that the text for the file uploader is `blue 02` (Color.Blue_02). 
Screen Shot 2022-12-21 at 11 59 48 AM
I would confirm with Anna before changing this to `blue 03`.

File: src/dex-ui-components/base/Inputs/FileUploader.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  fileUploaderStyles?: any;
+  disabled?: boolean;
+  onFileRead: (file: string, fileName: string) => void;
+  onFileReadFailed?: (reason: string) => void;
+}
+
+function FileUploader(props: FileUploaderProps) {
+  const { title, subTitle, onHoverTitle, fileUploaderStyles, disabled, onFileRead, onFileReadFailed } = props;
+
+  const onDrop = useCallback(
```
Comment (damiano-melcarne-lab49) [inline src/dex-ui-components/base/Inputs/FileUploader.tsx:33]: There's no need to use `useCallback` unless we're optimizing for a specific performance issue. It adds unnecessary bloat to the code and can even hurt performance (https://kentcdodds.com/blog/usememo-and-usecallback). You may have already noticed this but `useCallback` is unnecessarily being used in other components as well and will need to be refactored at a later point.

File: src/dex-ui-components/base/Inputs/FileUploader.tsx (hunk 1 above)
Comment (damiano-melcarne-lab49) [inline src/dex-ui-components/base/Inputs/FileUploader.tsx:11]: To stick with the ChakraUI styling config pattern, you can add these styles as a `variant` in the `CardStyles` object in **dex-ui-components/base/Card/styles.ts**.

`
export const CardStyles: ComponentStyleConfig = {
...
variants: {
"file-uploader": {
...styles
}
}
...
}
`

`
<Card {...getRootProps()} variant="file-uploader" style={{ ...fileUploaderStyles }}>
`

File: src/dex-ui-components/base/Inputs/FileUploader.tsx (hunk 3 above)
Comment (damiano-melcarne-lab49) [inline src/dex-ui-components/base/Inputs/FileUploader.tsx:33]: There's no need to use `useCallback` unless we're optimizing for a specific performance issue. It adds unnecessary bloat to the code and can even hurt performance (https://kentcdodds.com/blog/usememo-and-usecallback). You may have already noticed this but `useCallback` is unnecessarily being used in other components as well and will need to be refactored at a later point.

File: src/dex-ui-components/base/Inputs/FileUploader.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  return (
+    <Card {...getRootProps()} style={{ ...defaultFileUploaderStyles, ...fileUploaderStyles }}>
+      <input {...getInputProps()} />
+      {isDragActive ? (
+        <Text color={Color.Blue_03}>{onHoverTitle}</Text>
+      ) : (
+        <>
+          <Text color={Color.Blue_03}>{title}</Text>
+          <Text color={Color.Blue_03}>{subTitle}</Text>
```
Comment (damiano-melcarne-lab49) [inline src/dex-ui-components/base/Inputs/FileUploader.tsx:68]: Also, it looks like the wireframes specify that the text for the file uploader is `blue 02` (Color.Blue_02). 
Screen Shot 2022-12-21 at 11 59 48 AM
I would confirm with Anna before changing this to `blue 03`.

File: src/dex-ui-components/base/Inputs/FileUploader.tsx (hunk 2 above)
Comment (damiano-melcarne-lab49) [inline src/dex-ui-components/base/Inputs/FileUploader.tsx:67]: Use the `textStyle` variants to define text attributes. You can then override attributes like `color` using inline styles (like you are already doing). 

It looks like the wireframes specified this as `body-02` text which maps to `b2` in our code.
Screen Shot 2022-12-21 at 11 50 15 AM

`<Text textStyle="b2" color={Color.Blue_03}>{title}</Text>`

The list of available custom text variants extracted from the wireframes is in the **dex-ui-components/base/Text/styles/TextStyles.ts** file.

Note that Typography defined in the wireframes should but may not always map one-to-one with the text variants described in our code. If you run into any discrepancies, you can reach out to myself or Anna.

File: src/dex-ui-components/base/Inputs/FileUploader.tsx (hunk 1 above)
Comment (roshan-bisht-lab49) [reply src/dex-ui-components/base/Inputs/FileUploader.tsx]: Done, I had it before but somehow the UI was giving correct result using variant though I have made the change but we can discuss this, may be I am missing something. Today I can check again.

File: src/dex-ui-components/base/Inputs/FileUploader.tsx (hunk 3 above)
Comment (roshan-bisht-lab49) [reply src/dex-ui-components/base/Inputs/FileUploader.tsx]: Yes agree, I have removed it.

File: src/dex-ui-components/base/Inputs/FileUploader.tsx (hunk 4 above)
Comment (roshan-bisht-lab49) [reply src/dex-ui-components/base/Inputs/FileUploader.tsx]: For new Color, the available color was not able in our color pallete so I added a new one, I can check with Anna today and change if required.

File: src/dex-ui-components/base/Inputs/FileUploader.tsx (hunk 2 above)
Comment (roshan-bisht-lab49) [reply src/dex-ui-components/base/Inputs/FileUploader.tsx]: Done.

**Review by damiano-melcarne-lab49 — APPROVED** [review]

File: src/dex-ui-components/base/Inputs/FileUploader.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+  const { getRootProps, getInputProps, isDragActive } = useDropzone({
+    onDrop,
+    maxFiles: 1,
+    accept: { "application/json": [".json"] },
+    disabled: disabled,
+  });
+
+  return (
+    <Card id={id} {...getRootProps()} variant="file-uploader" style={{ ...fileUploaderStyles }} cursor="pointer">
```
Comment (damiano-melcarne-lab49) [inline src/dex-ui-components/base/Inputs/FileUploader.tsx:48]: `cursor` can be added in the variant styles in the future.

```
"file-uploader": {
   ....
   cursor:"pointer"
}
```

**Review by damiano-melcarne-lab49 — APPROVED** [review]

File: src/dex-ui-components/base/Inputs/FileUploader.tsx (hunk 5 above)
Comment (roshan-bisht-lab49) [reply src/dex-ui-components/base/Inputs/FileUploader.tsx:48]: Got it, will fix this in next PR (81).

**Pull request merged**